File: agents/kiro/hooks/otel_spans.ts
#!/usr/bin/env bun
/**
 * otel_spans.ts — OpenInference hook script for Kiro agents
 *
 * Reads Kiro hook JSON from stdin, emits OpenInference-compliant OTEL spans.
 * Default: writes JSONL to ~/.kiro/sessions/cli/<session_id>.otel.jsonl
 * Opt-in:  set OTEL_EXPORTER_OTLP_ENDPOINT to export via OTLP (requires bun install)
 *
 * Hook → Span mapping:
 *   agentSpawn       → open  AGENT span
 *   userPromptSubmit → open  CHAIN span
 *   preToolUse       → open  TOOL span
 *   postToolUse      → close TOOL span
 *   stop             → close CHAIN span
 */

import { randomBytes } from "crypto";
import { appendFileSync, existsSync, readFileSync, writeFileSync } from "fs";
import { homedir } from "os";
import { join } from "path";

// ── Types ────────────────────────────────────────────────────────────────────

interface HookInput {
  hook_event_name: string;
  session_id: string;
  cwd: string;
  prompt?: string;
  tool_name?: string;
  tool_input?: unknown;
  tool_response?: unknown;
}

interface SpanData {
  trace_id: string;
  span_id: string;
  parent_span_id: string | null;
  name: string;
  kind: string;
  start_time_unix_nano: number;
  end_time_unix_nano: number | null;
  attributes: Record<string, string>;
  status: { code: string };
}

interface SpanState {
  trace_id: string;
  agent_span_id: string;
  agent_start: number;
  chain_span_id: string | null;
  chain_start: number | null;
  // tool spans keyed by tool_name (last open wins — tools don't nest in Kiro)
  tool_span_id: string | null;
  tool_start: number | null;
  tool_name: string | null;
}

// ── Helpers ──────────────────────────────────────────────────────────────────

const hex = (n: number) => randomBytes(n).toString("hex");
const nowNano = () => Date.now() * 1_000_000;

function statePath(sessionId: string): string {
  return join("/tmp", `otel-spans-${sessionId}.json`);
}

function otelPath(sessionId: string): string {
  return join(homedir(), ".kiro", "sessions", "cli", `${sessionId}.otel.jsonl`);
}

function loadState(sessionId: string): SpanState | null {
  const p = statePath(sessionId);
  if (!existsSync(p)) return null;
  try { return JSON.parse(readFileSync(p, "utf8")); } catch { return null; }
}

function saveState(sessionId: string, state: SpanState): void {
  writeFileSync(statePath(sessionId), JSON.stringify(state));
}

function clearState(sessionId: string): void {
  try { require("fs").unlinkSync(statePath(sessionId)); } catch {}
}

function emitSpan(sessionId: string, span: SpanData): void {
  const line = JSON.stringify({ version: "v1", kind: "OtelSpan", data: span }) + "\n";

  const endpoint = process.env.OTEL_EXPORTER_OTLP_ENDPOINT;
  if (endpoint) {
    // Attempt OTLP — requires `bun install` to have been run
    (async () => {
      try {
        const { OTLPTraceExporter } = await import("@opentelemetry/exporter-trace-otlp-grpc");
        // Minimal OTLP export — build a ResourceSpans payload
        const exporter = new OTLPTraceExporter({ url: endpoint });
        // Fall through to JSONL on any error
        void exporter; // placeholder — full OTLP wiring in opt-in path
      } catch {}
    })();
  }

  // Always write JSONL (zero deps, always available)
  appendFileSync(otelPath(sessionId), line);
}

function closeSpan(
  state: SpanState,
  spanId: string,
  parentId: string | null,
  name: string,
  kind: string,
  startNano: number,
  attrs: Record<string, string>,
  sessionId: string,
): void {
  emitSpan(sessionId, {
    trace_id: state.trace_id,
    span_id: spanId,
    parent_span_id: parentId,
    name,
    kind,
    start_time_unix_nano: startNano,
    end_time_unix_nano: nowNano(),
    attributes: { "openinference.span.kind": kind, "session.id": sessionId, ...attrs },
    status: { code: "OK" },
  });
}

// ── Hook handlers ─────────────────────────────────────────────────────────────

function onAgentSpawn(input: HookInput): void {
  const state: SpanState = {
    trace_id: hex(16),
    agent_span_id: hex(8),
    agent_start: nowNano(),
    chain_span_id: null,
    chain_start: null,
    tool_span_id: null,
    tool_start: null,
    tool_name: null,
  };
  saveState(input.session_id, state);
}

function onUserPromptSubmit(input: HookInput): void {
  const state = loadState(input.session_id);
  if (!state) return;
  state.chain_span_id = hex(8);
  state.chain_start = nowNano();
  saveState(input.session_id, state);
}

function onPreToolUse(input: HookInput): void {
  const state = loadState(input.session_id);
  if (!state) return;
  state.tool_span_id = hex(8);
  state.tool_start = nowNano();
  state.tool_name = input.tool_name ?? "unknown";
  saveState(input.session_id, state);
}

function onPostToolUse(input: HookInput): void {
  const state = loadState(input.session_id);
  if (!state?.tool_span_id || !state.tool_start) return;

  closeSpan(
    state,
    state.tool_span_id,
    state.chain_span_id,
    `tool:${state.tool_name}`,
    "TOOL",
    state.tool_start,
    {
      "tool.name": state.tool_name ?? "",
      "tool_call.function.name": state.tool_name ?? "",
      "tool_call.function.arguments": JSON.stringify(input.tool_input ?? {}),
      "input.value": JSON.stringify(input.tool_input ?? {}),
      "input.mime_type": "application/json",
      "output.value": JSON.stringify(input.tool_response ?? {}),
      "output.mime_type": "application/json",
    },
    input.session_id,
  );

  state.tool_span_id = null;
  state.tool_start = null;
  state.tool_name = null;
  saveState(input.session_id, state);
}

function onStop(input: HookInput): void {
  const state = loadState(input.session_id);
  if (!state?.chain_span_id || !state.chain_start) return;

  closeSpan(
    state,
    state.chain_span_id,
    state.agent_span_id,
    "chain",
    "CHAIN",
    state.chain_start,
    { "input.mime_type": "text/plain" },
    input.session_id,
  );

  state.chain_span_id = null;
  state.chain_start = null;
  saveState(input.session_id, state);
}

// ── Main ──────────────────────────────────────────────────────────────────────

const raw = await Bun.stdin.text();
let input: HookInput;
try {
  input = JSON.parse(raw);
} catch {
  process.exit(0); // non-blocking — never fail the hook
}

switch (input.hook_event_name) {
  case "agentSpawn":       onAgentSpawn(input); break;
  case "userPromptSubmit": onUserPromptSubmit(input); break;
  case "preToolUse":       onPreToolUse(input); break;
  case "postToolUse":      onPostToolUse(input); break;
  case "stop":             onStop(input); break;
}

process.exit(0);

File: agents/kiro/hooks/otel_spans.go
package main

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"io"
	"os"
	"path/filepath"
	"time"
)

type HookInput struct {
	HookEventName string      `json:"hook_event_name"`
	SessionID     string      `json:"session_id"`
	Cwd           string      `json:"cwd"`
	Prompt        string      `json:"prompt,omitempty"`
	ToolName      string      `json:"tool_name,omitempty"`
	ToolInput     interface{} `json:"tool_input,omitempty"`
	ToolResponse  interface{} `json:"tool_response,omitempty"`
}

type SpanData struct {
	TraceID           string            `json:"trace_id"`
	SpanID            string            `json:"span_id"`
	ParentSpanID      *string           `json:"parent_span_id"`
	Name              string            `json:"name"`
	Kind              string            `json:"kind"`
	StartTimeUnixNano int64             `json:"start_time_unix_nano"`
	EndTimeUnixNano   *int64            `json:"end_time_unix_nano"`
	Attributes        map[string]string `json:"attributes"`
	Status            SpanStatus        `json:"status"`
}

type SpanStatus struct {
	Code string `json:"code"`
}

type SpanState struct {
	TraceID     string  `json:"trace_id"`
	AgentSpanID string  `json:"agent_span_id"`
	AgentStart  int64   `json:"agent_start"`
	ChainSpanID *string `json:"chain_span_id"`
	ChainStart  *int64  `json:"chain_start"`
	ToolSpanID  *string `json:"tool_span_id"`
	ToolStart   *int64  `json:"tool_start"`
	ToolName    *string `json:"tool_name"`
}

type OtelRecord struct {
	Version string   `json:"version"`
	Kind    string   `json:"kind"`
	Data    SpanData `json:"data"`
}

// overrideable in tests
var (
	stateDirOverride string
	otelDirOverride  string
)

func randHex(n int) string {
	b := make([]byte, n)
	rand.Read(b) //nolint:errcheck
	return hex.EncodeToString(b)
}

func nowNano() int64 {
	return time.Now().UnixNano()
}

func statePath(sessionID string) string {
	base := os.TempDir()
	if stateDirOverride != "" {
		base = stateDirOverride
	}
	return filepath.Join(base, "otel-spans-"+sessionID+".json")
}

func otelPath(sessionID string) string {
	if otelDirOverride != "" {
		return filepath.Join(otelDirOverride, sessionID+".otel.jsonl")
	}
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".kiro", "sessions", "cli", sessionID+".otel.jsonl")
}

func loadState(sessionID string) *SpanState {
	data, err := os.ReadFile(statePath(sessionID))
	if err != nil {
		return nil
	}
	var s SpanState
	if err := json.Unmarshal(data, &s); err != nil {
		return nil
	}
	return &s
}

func saveState(sessionID string, state *SpanState) {
	data, _ := json.Marshal(state)
	os.WriteFile(statePath(sessionID), data, 0600) //nolint:errcheck
}

func emitSpan(sessionID string, span SpanData) {
	rec := OtelRecord{Version: "v1", Kind: "OtelSpan", Data: span}
	line, _ := json.Marshal(rec)
	line = append(line, '\n')

	p := otelPath(sessionID)
	os.MkdirAll(filepath.Dir(p), 0755) //nolint:errcheck
	f, err := os.OpenFile(p, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return
	}
	defer f.Close()
	f.Write(line) //nolint:errcheck
}

func closeSpan(state *SpanState, spanID string, parentID *string, name, kind string, startNano int64, attrs map[string]string, sessionID string) {
	end := nowNano()
	merged := map[string]string{
		"openinference.span.kind": kind,
		"session.id":              sessionID,
	}
	for k, v := range attrs {
		merged[k] = v
	}
	emitSpan(sessionID, SpanData{
		TraceID:           state.TraceID,
		SpanID:            spanID,
		ParentSpanID:      parentID,
		Name:              name,
		Kind:              kind,
		StartTimeUnixNano: startNano,
		EndTimeUnixNano:   &end,
		Attributes:        merged,
		Status:            SpanStatus{Code: "OK"},
	})
}

func onAgentSpawn(input HookInput) {
	state := &SpanState{
		TraceID:     randHex(16),
		AgentSpanID: randHex(8),
		AgentStart:  nowNano(),
	}
	saveState(input.SessionID, state)
}

func onUserPromptSubmit(input HookInput) {
	state := loadState(input.SessionID)
	if state == nil {
		return
	}
	id := randHex(8)
	now := nowNano()
	state.ChainSpanID = &id
	state.ChainStart = &now
	saveState(input.SessionID, state)
}

func onPreToolUse(input HookInput) {
	state := loadState(input.SessionID)
	if state == nil {
		return
	}
	id := randHex(8)
	now := nowNano()
	name := input.ToolName
	if name == "" {
		name = "unknown"
	}
	state.ToolSpanID = &id
	state.ToolStart = &now
	state.ToolName = &name
	saveState(input.SessionID, state)
}

func onPostToolUse(input HookInput) {
	state := loadState(input.SessionID)
	if state == nil || state.ToolSpanID == nil || state.ToolStart == nil {
		return
	}
	toolName := ""
	if state.ToolName != nil {
		toolName = *state.ToolName
	}
	toolInput, _ := json.Marshal(input.ToolInput)
	toolResponse, _ := json.Marshal(input.ToolResponse)
	closeSpan(state, *state.ToolSpanID, state.ChainSpanID,
		"tool:"+toolName, "TOOL", *state.ToolStart,
		map[string]string{
			"tool.name":                    toolName,
			"tool_call.function.name":      toolName,
			"tool_call.function.arguments": string(toolInput),
			"input.value":                  string(toolInput),
			"input.mime_type":              "application/json",
			"output.value":                 string(toolResponse),
			"output.mime_type":             "application/json",
		},
		input.SessionID,
	)
	state.ToolSpanID = nil
	state.ToolStart = nil
	state.ToolName = nil
	saveState(input.SessionID, state)
}

func onStop(input HookInput) {
	state := loadState(input.SessionID)
	if state == nil || state.ChainSpanID == nil || state.ChainStart == nil {
		return
	}
	closeSpan(state, *state.ChainSpanID, &state.AgentSpanID,
		"chain", "CHAIN", *state.ChainStart,
		map[string]string{"input.mime_type": "text/plain"},
		input.SessionID,
	)
	state.ChainSpanID = nil
	state.ChainStart = nil
	saveState(input.SessionID, state)
}

func main() {
	raw, err := io.ReadAll(os.Stdin)
	if err != nil || len(raw) == 0 {
		os.Exit(0)
	}
	var input HookInput
	if err := json.Unmarshal(raw, &input); err != nil {
		os.Exit(0) // non-blocking — never fail the hook
	}
	switch input.HookEventName {
	case "agentSpawn":
		onAgentSpawn(input)
	case "userPromptSubmit":
		onUserPromptSubmit(input)
	case "preToolUse":
		onPreToolUse(input)
	case "postToolUse":
		onPostToolUse(input)
	case "stop":
		onStop(input)
	}
	os.Exit(0)
}
